libutils/json.c

```diff
@@ -296,31 +298,24 @@ static int JsonObjectCompare(
     int ret = JsonLength(a) - JsonLength(b);
     if (ret != 0)
     {
+        Log(LOG_LEVEL_DEBUG, "JsonObjectCompare() fails, length differs by %d", ret);
         return ret;
     }
 
-    JsonIterator iter_a = JsonIteratorInit(a);
-    JsonIterator iter_b = JsonIteratorInit(a);
+    JsonIterator iter = JsonIteratorInit(a);
 
-    for (size_t i = 0; i < JsonLength(a); i++)
+    while (JsonIteratorHasMore(&iter))
     {
-        const JsonElement *child_a = JsonIteratorNextValue(&iter_a);
-        const JsonElement *child_b = JsonIteratorNextValue(&iter_b);
-
-        const char *const key_a = JsonIteratorCurrentKey(&iter_a);
-        const char *const key_b = JsonIteratorCurrentKey(&iter_b);
-
-        ret = strcmp(key_a, key_b);
-        if (ret != 0)
-        {
-            return ret;
-        }
+        const char *const key = JsonIteratorNextKey(&iter);
+        const JsonElement *const child_a = JsonObjectGet(a, key);
+        const JsonElement *const child_b = JsonObjectGet(b, key);
 
-        ret = JsonCompare(child_a, child_b);
-        if (ret != 0)
+        if (child_b == NULL || JsonCompare(child_a, child_b))
         {
-            return ret;
+            Log(LOG_LEVEL_DEBUG, "JsonObjectCompare() fails for key '%s', children not equal", key);
+            return 1;
         }
+        return 0;
```

Comment:
```suggestion
        return 0; // equal, like strcmp()
```

Reply:
This return 0 inside the loop is actually rather wrong... it would cause us to only check the first child element of each object I think. I will add a test to check the code as-is.

Reply:
fixed, sorry about the json formatting at the end (hard to read but there)

```
   debug: JsonCompare() fails, primitive 'due' not equal to 'dva'
   debug: JsonObjectCompare() fails for key 'two', children are not equal
   debug: JsonObjectCompare() fails for key 'badchild', children are not equal
   debug: JsonObjectCompare() fails for key 'cfbs:delete_files.filenames_bad_two', children are not equal
   debug: JsonObjectCompare() fails for key 'variables_bad_two', children are not equal
merged json  : {"variables_bad_two":{"cfbs:delete_files.filenames_bad_two":{"badchild":{"one":"uno","two":"due"},"value":["/tmp/foo","/tmp/bar","/tmp/bar","/tmp/baz"]}}}
expected json: {"variables_bad_two":{"cfbs:delete_files.filenames_bad_two":{"badchild":{"one":"uno","two":"dva"},"value":["/tmp/foo","/tmp/bar","/tmp/bar","/tmp/baz"]}}}
```